### PR TITLE
Adaptations neededed for gas-water system

### PR DESCRIPTION
**Pull request by plgbrts**
Material laws needed for two-phase water-gas system.
Further work is required and ongoing on opm-models  and opm-simulators.

### CODE REVIEW

**Comment by blattms** [comment]
jenkins build this please

File: opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp [hunk 1]
```diff
@@ -868,6 +947,51 @@ class EclMaterialLawManager
...
+        dest[satRegionIdx] = std::make_shared<GasWaterEffectiveTwoPhaseParams>();
+
+        auto& effParams = *dest[satRegionIdx];
+
+        const auto& tableManager = eclState.getTableManager();
+
+        switch (eclState.runspec().saturationFunctionControls().family()) {
+        case SatFuncControls::KeywordFamily::Family_I:
+        {
+            // No Family I applicable in this case
```
Comment (bska) [inline opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp:968]: Doesn't `SLGOF` model this case?

File: opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp [hunk 2]
```diff
@@ -868,6 +947,51 @@ class EclMaterialLawManager
...
+        case SatFuncControls::KeywordFamily::Family_II:
+        {
+            //Todo: allow also for Sgwfn table input as alternative to Sgfn and Swfn table input
+            const SgfnTable& sgfnTable = tableManager.getSgfnTables().getTable<SgfnTable>( satRegionIdx );
+            const SwfnTable& swfnTable = tableManager.getSwfnTables().getTable<SwfnTable>( satRegionIdx );
+
+            std::vector<double> SwColumn = swfnTable.getColumn("SW").vectorCopy();
+            
+            effParams.setKrwSamples(SwColumn, swfnTable.getColumn("KRW").vectorCopy());
+            effParams.setKrnSamples(SwColumn, sgfnTable.getColumn("KRG").vectorCopy());
```
Comment (bska) [inline opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp:981]: Please use [`normalizeKrValues()`](https://github.com/OPM/opm-material/blob/ae3791b2b013e124e122d4c6d8274b895828970d/opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp#L993-L1004) here to account for `TOLCRIT`.  See PR #439 (commit bc38dcb189158e6575e2cb694093d32c5f74f772) for additional details.

**Review by bska — COMMENTED** [review]
It looks like you're missing the end-point scaling feature for gas-water.  Do you plan to add that in a future PR?

File: opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp (hunk 1 above)
Comment (plgbrts) [reply opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp:971]: I believe not as SLGOF  does not allow for specification of gas-water capillary pressure.  For gas-water systems also eclipse requires specification of saturation functions for gas-water only by family II keywords

File: opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp (hunk 2 above)
Comment (plgbrts) [reply opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp]: > It looks like you're missing the end-point scaling feature for gas-water. Do you plan to add that in a future PR?

That I will do in a future PR.

**Comment by bska** [comment]
jenkins build this please

File: opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp (hunk 1 above)
Comment (bska) [reply opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp:971]: >> Doesn't `SLGOF` model this case?
>
> I believe not as `SLGOF` does not allow for specification of gas-water capillary pressure.

Yes you're right, sorry about the confusion.

**Review by bska — COMMENTED** [review]
Sorry about the delayed response.  I think this is a good start and I'd like to get this merged into the master branch.  We do however need to do some more work here before we're able to do so.  The current version does not compile on the Jenkins [CI system](https://ci.opm-project.org/job/opm-material-PR-builder/135/console) when we include the unit tests.

We do, for instance, get diagnostic messages of the form
```
/var/lib/jenkins/post-builder/workspace/opm-material-PR-builder/tests/test_fluidmatrixinteractions.cpp:344:79: error: wrong number of template arguments (3, should be at least 4)
                                          /*OilWaterMaterial=*/TwoPhaseMaterial> MaterialLaw;
                                                                               ^
In file included from /var/lib/jenkins/post-builder/workspace/opm-material-PR-builder/tests/test_fluidmatrixinteractions.cpp:52:
/var/lib/jenkins/post-builder/workspace/opm-material-PR-builder/opm/material/fluidmatrixinteractions/EclTwoPhaseMaterial.hpp:56:7: note: provided for 'template<class TraitsT, class GasOilMaterialLawT, class OilWaterMaterialLawT, class GasWaterMaterialLawT, class ParamsT> class Opm::EclTwoPhaseMaterial'
 class EclTwoPhaseMaterial : public TraitsT
       ^~~~~~~~~~~~~~~~~~~
/var/lib/jenkins/post-builder/workspace/opm-material-PR-builder/tests/test_fluidmatrixinteractions.cpp:353:82: error: wrong number of template arguments (3, should be at least 4)
                                             /*OilWaterMaterial=*/TwoPhaseMaterial> MaterialLaw;
                                                                                  ^
```
when testing the API of a [two-phase material law](https://github.com/OPM/opm-material/blob/a1dd037984b4f7ee2884d9fa3d21219da13161f1/tests/test_fluidmatrixinteractions.cpp#L342-L344) built on the Brooks-Corey relation.

There may also be other changes needed in downstream repositories such as opm-simulators but those are not exposed until we're at least able to compile this PR in isolation on the CI system.

**Comment by plgbrts** [comment]
> Sorry about the delayed response. I think this is a good start and I'd like to get this merged into the master branch. We do however need to do some more work here before we're able to do so. The current version does not compile on the Jenkins [CI system](https://ci.opm-project.org/job/opm-material-PR-builder/135/console) when we include the unit tests.
> 
> We do, for instance, get diagnostic messages of the form
> 
> ```
> /var/lib/jenkins/post-builder/workspace/opm-material-PR-builder/tests/test_fluidmatrixinteractions.cpp:344:79: error: wrong number of template arguments (3, should be at least 4)
>                                           /*OilWaterMaterial=*/TwoPhaseMaterial> MaterialLaw;
>                                                                                ^
> In file included from /var/lib/jenkins/post-builder/workspace/opm-material-PR-builder/tests/test_fluidmatrixinteractions.cpp:52:
> /var/lib/jenkins/post-builder/workspace/opm-material-PR-builder/opm/material/fluidmatrixinteractions/EclTwoPhaseMaterial.hpp:56:7: note: provided for 'template<class TraitsT, class GasOilMaterialLawT, class OilWaterMaterialLawT, class GasWaterMaterialLawT, class ParamsT> class Opm::EclTwoPhaseMaterial'
>  class EclTwoPhaseMaterial : public TraitsT
>        ^~~~~~~~~~~~~~~~~~~
> /var/lib/jenkins/post-builder/workspace/opm-material-PR-builder/tests/test_fluidmatrixinteractions.cpp:353:82: error: wrong number of template arguments (3, should be at least 4)
>                                              /*OilWaterMaterial=*/TwoPhaseMaterial> MaterialLaw;
>                                                                                   ^
> ```
> 
> when testing the API of a [two-phase material law](https://github.com/OPM/opm-material/blob/a1dd037984b4f7ee2884d9fa3d21219da13161f1/tests/test_fluidmatrixinteractions.cpp#L342-L344) built on the Brooks-Corey relation.
> 
> There may also be other changes needed in downstream repositories such as opm-simulators but those are not exposed until we're at least able to compile this PR in isolation on the CI system.

I fixed the test file. Hopefully the PR can be compiled now including the tests.

**Comment by bska** [comment]
> I fixed the test file.

Very good, much appreciated.  I'll run a new build check.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
> I'll run a new build check.

Sadly, that check turned up a number of [blocking issues](https://ci.opm-project.org/job/opm-material-PR-builder/136/testReport/).  We need to investigate and resolve those before we can merge this into the master branch.

**Comment by bska** [comment]
> > I'll run a new build check.
> 
> Sadly, that check turned up a number of [blocking issues](https://ci.opm-project.org/job/opm-material-PR-builder/136/testReport/).

I'll run the build check once again.  Some of those regressions are probably caused by an update to the reference solutions that were added while we were running.  The new build check should provide an accurate overview of where we need to focus our efforts.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
> the only thing missing the update was SPE1CASE1_WATER, but sure, may as well get that one eliminated from the list.

Fair enough, but we also get the following [diagnostic message](https://ci.opm-project.org/job/opm-material-PR-builder/137/consoleFull).  Any idea what that's all about, @akva2?

```
/var/lib/jenkins/workspace/opm-material-PR-builder/deps/opm-simulators/jenkins/run-spe.sh: line 19: /var/lib/jenkins/workspace/opm-material-PR-builder/mpi/build-opm-simulators/bin/flow_legacy: No such file or directory
```

**Comment by akva2** [comment]
ehm, my bad. i commented while it was in the queue. there is a bug in the jenkins PR builder that see the last comment, not the comment that actually triggered. so it found saw
```
ghprobCommentBody=the only thing missing the update was SPE1CASE1_WATER, but sure, may as well get that one 
eliminated from the list.
```
that is, it saw 'spe' in the trigger, and thus wanted to run the spe extra tests which have not been updated since the flow_legacy days. it can be safely ignored.

**Comment by bska** [comment]
> I commented while it was in the queue [...] so [the jenkins trigger] found [spe]

Ah, I see.  Very well.  I'll just run this one more time to get the final impression of what is failing.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
> I'll just run this one more time to get the final impression of what is failing.

There are 36 failing regression tests in the latest [test report](https://ci.opm-project.org/job/opm-material-PR-builder/138/testReport/).  Some of those are segmentation faults that look like index-out-of-bounds errors.  I have not looked at all of them however.  Still, these results do mean that the current PR is not ready for merging into the master branch and that more work is needed.

I suggest you start by running the [`eclmateriallawmanager`](https://ci.opm-project.org/job/opm-material-PR-builder/138/testReport/(root)/mpi/eclmateriallawmanager/) test on your local machine to see if you're able to reproduce the symptoms in a controlled environment and then resolve this segmentation fault first.  Please don't hesitate to ask if you have any questions.

**Comment by plgbrts** [comment]
This update contains a fix to a avoid reading gas-water parameters when hasOil is true.  This fix should solve for the failed tests.

I checked this modification (only) for two decks (model2/8_MULTIPLY_TRANXYZ_MODEL2.DATA and /spe5/SPE5CASE1.DATA) that failed earlier but do run now without errors.

I was wondering how I could improve local testing before doing a PR (e.g. using testing scripts/tools to run local regression tests). Any advice will be appreciated.

**Comment by bska** [comment]
jenkins build this please

**Comment by alfbr** [comment]
> I was wondering how I could improve local testing before doing a PR (e.g. using testing scripts/tools to run local regression tests). Any advice will be appreciated.

You can run "make test" to have them all run locally. You can also run them one by one manually calling compareECL similarly as done in the cmake set-up file in opm-simulators. We should probably document this on the wiki.

**Comment by bska** [comment]
> > was wondering how I could improve local testing before doing a PR (e.g. using testing scripts/tools to run local regression tests). Any advice will be appreciated.
> 
> You can run "make test" to have them all run locally.

You should be aware, though, that your mileage may vary with this approach, especially in opm-simulators.  I personally don't get any useful information if I "just" run `make test` in opm-simulators.  The reference solutions in opm-tests are usually too different from those I calculate locally, so I typically have a separate build of all repositories which I keep up-to-date with the current master sources but otherwise unchanged (i.e., none of my own in-development code changes here).  I then run
```
ctest --parallel 7
```
('7' happens to fit my local machine&mdash;adjust as needed) in **both** the master and development build directories and then generate a comparison script that uses `compareECL` as follows
```sh
ref=path/to/master/build/opm-simulators
cand=path/to/development/build/opm-simulators
cd "${ref}"
{ find . -name '*.UNRST' | \
    while read r; do
        c="${r%*.UNRST}"
        echo "${ref}/../opm-common/bin/compareECL" -a "'${ref}/${c}'" "'${cand}/${c}'" 2.0e-2 1.0e-5
    done; } >/tmp/x/compare.sh
```
Then I can run the comparison script `/tmp/x/compare.sh`:
```sh
time { sh /tmp/x/compare.sh >compare.out 2>&1; }
```
It is sometimes helpful to extract the list of cases from that so I often do this in addition to the above
```sh
sed -ne "s/^Comparing '\([^']\{1,\}\)' to '\([^']\{1,\}\)'/\1 \2/p" ./compare.out >cases.txt
```
which gives the "base names" of the result files.  The manual part of this is then to look through "compare.out" to determine if any regression tests have actually failed and then to possibly go on to look at the individual failures if the differences are large enough.  The command
```sh
grep exhibit ./compare.out
```
will help identify if any regression tests have failed at all.  If there's no output from that command then you can typically assume that your changes will be okay from the point of view of not breaking any regression tests.

**Comment by bska** [comment]
Thanks a lot for the updates.  This is much better than the previous iteration.  As far as I can tell, there is now just a single failing regression case, namely the single-phase water case [flow_onephase_SPE1CASE1_WATER](https://ci.opm-project.org/job/opm-material-PR-builder/140/testReport/(root)/mpi/compareECLFiles_flow_onephase_SPE1CASE1_WATER/).  The diagnostic message is
```
Program threw an exception: Could not initialize the problem: No valid saturation keyword family specified
```
If you fix that then I'll merge the PR.

**Comment by blattms** [comment]
> ```
> ctest --parallel 7
> ```

Does parallel testing work now in general? The last time I tried, it did fail, because some of the tests depend on results from others and this dependency cannot be communicated to CMake. But I made have tried more than 7 threads.

**Comment by akva2** [comment]
we have been running 4 tests in parallel on jenkins for several years now. it works fine.

**Comment by plgbrts** [comment]
This small update should fix the failing of the one phase test case SPE1CASE1_WATER.DATA.
Many thanks for all the good advice about testing!

File: opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp [hunk 3]
```diff
@@ -868,6 +947,54 @@ class EclMaterialLawManager
...
+        const auto tolcrit = eclState.runspec().saturationFunctionControls()
+            .minimumRelpermMobilityThreshold();
+
+        const auto& tableManager = eclState.getTableManager();
+
+        switch (eclState.runspec().saturationFunctionControls().family()) {
+        case SatFuncControls::KeywordFamily::Family_I:
+        {
+            // No Family I applicable in this case
+            break;
```
Comment (totto82) [inline opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp:972]: Does it make sense to throw here?

**Comment by totto82** [comment]
jenkins build this please

File: opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp (hunk 3 above)
Comment (plgbrts) [reply opm/material/fluidmatrixinteractions/EclMaterialLawManager.hpp]: I agree, that does make sense. I included a throw

**Comment by bska** [comment]
Thanks a lot for the updates.  I'll run a final build check and then merge when it comes back green.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
Build check is green.  I'll merge into master.  Thanks a lot for the contribution.

**Pull request merged**